File: src/main/java/me/testaccount666/serversystem/clickablesigns/executables/kit/ConfiguratorKitSign.java
package me.testaccount666.serversystem.clickablesigns.executables.kit;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.clickablesigns.AbstractSignConfigurator;
import me.testaccount666.serversystem.clickablesigns.SignType;
import me.testaccount666.serversystem.commands.executables.kit.manager.KitManager;
import me.testaccount666.serversystem.userdata.User;
import me.testaccount666.serversystem.utils.ChatColor;
import me.testaccount666.serversystem.utils.ComponentColor;
import org.bukkit.block.Sign;
import org.bukkit.block.sign.Side;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;

import static me.testaccount666.serversystem.utils.MessageBuilder.sign;

public class ConfiguratorKitSign extends AbstractSignConfigurator {

    @Override
    protected String getCreatePermissionNode() {
        return "ClickableSigns.Kit.Create";
    }

    @Override
    protected SignType getSignType() {
        return SignType.KIT;
    }

    @Override
    protected boolean validateConfiguration(User user, Sign sign, YamlConfiguration config) {
        var kitManager = ServerSystem.Instance.getRegistry().getService(KitManager.class);
        if (kitManager == null) {
            sign("Kit.NoKitManager", user).build();
            return false;
        }

        var front = sign.getSide(Side.FRONT);
        var kitName = ComponentColor.componentToString(front.line(1));
        if (kitName.isEmpty()) {
            sign("Kit.NoKitSpecified", user).build();
            return false;
        }

        if (!kitManager.kitExists(kitName.toLowerCase())) {
            sign("Kit.KitNotFound", user)
                    .postModifier(message -> message.replace("<KIT>", kitName)).build();
            return false;
        }

        front.line(0, ComponentColor.translateToComponent(SignType.KIT.signName()));
        front.line(1, ComponentColor.translateToComponent("&2${kitName}"));
        var back = sign.getSide(Side.BACK);
        for (var index = 0; index < 4; index++) back.line(index, front.line(index));
        sign.update();
        return true;
    }

    @Override
    protected void addSignSpecificConfiguration(User user, Sign sign, FileConfiguration config) {
        var kitName = sign.getSide(Side.FRONT).getLine(1);
        kitName = ChatColor.stripColor(kitName);
        config.set("KitName", kitName);
    }

    @Override
    protected String getSuccessMessageKey() {
        return "Kit.Created";
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/clickablesigns/executables/kit/ConfiguratorKitSign.kt
package me.testaccount666.serversystem.clickablesigns.executables.kit

import me.testaccount666.paperktx.colors.ChatColor.Companion.stripColor
import me.testaccount666.paperktx.extensions.ComponentExtensions.asComponent
import me.testaccount666.paperktx.extensions.ComponentExtensions.asString
import me.testaccount666.serversystem.clickablesigns.AbstractSignConfigurator
import me.testaccount666.serversystem.clickablesigns.SignType
import me.testaccount666.serversystem.commands.executables.kit.manager.KitManager
import me.testaccount666.serversystem.extensions.getService
import me.testaccount666.serversystem.extensions.signMsg
import me.testaccount666.serversystem.userdata.User
import org.bukkit.block.Sign
import org.bukkit.block.sign.Side
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.configuration.file.YamlConfiguration

class ConfiguratorKitSign : AbstractSignConfigurator() {
    override val createPermissionNode = "ClickableSigns.Kit.Create"
    override val successMessageKey = "Kit.Created"
    override val signType = SignType.KIT

    override fun validateConfiguration(user: User, sign: Sign, config: YamlConfiguration): Boolean {
        val front = sign.getSide(Side.FRONT)
        val kitName = front.line(1).asString().takeUnless { it.isEmpty() } ?: run {
            user.signMsg("Kit.NoKitSpecified")
            return false
        }

        if (!getService<KitManager>().kitExists(kitName)) {
            user.signMsg("Kit.KitNotFound") {
                postModifier { it.replace("<KIT>", kitName) }
            }
            return false
        }

        front.line(0, SignType.KIT.signName.asComponent())
        front.line(1, "&2${kitName}".asComponent())
        val back = sign.getSide(Side.BACK)
        for (index in 0..3) back.line(index, front.line(index))
        sign.update()
        return true
    }

    override fun addSignSpecificConfiguration(user: User, sign: Sign, config: FileConfiguration) {
        var kitName = sign.getSide(Side.FRONT).getLine(1)
        kitName = stripColor(kitName)
        config["KitName"] = kitName
    }
}